gps/pkgtree/pkgtree.go

```diff
@@ -607,38 +606,28 @@ func (t PackageTree) ToReachMap(main, tests, backprop bool, ignore *IgnoredRules
 // This is really only useful as a defensive measure to prevent external state
 // mutations.
 func (t PackageTree) Copy() PackageTree {
-	t2 := PackageTree{
+	return PackageTree{
 		ImportRoot: t.ImportRoot,
-		Packages:   make(map[string]PackageOrErr, len(t.Packages)),
+		Packages:   CopyPackages(t.Packages, nil),
 	}
+}
 
+// CopyPackages returns a deep copy of p, optionally modifying the entries with fn.
+func CopyPackages(p map[string]PackageOrErr, fn func(string, PackageOrErr) (string, PackageOrErr)) map[string]PackageOrErr {
+	p2 := make(map[string]PackageOrErr, len(p))
 	// Walk through and count up the total number of string slice elements we'll
 	// need, then allocate them all at once.
 	strcount := 0
-	for _, poe := range t.Packages {
+	for _, poe := range p {
 		strcount = strcount + len(poe.P.Imports) + len(poe.P.TestImports)
 	}
 	pool := make([]string, strcount)
 
-	for path, poe := range t.Packages {
+	for path, poe := range p {
 		var poe2 PackageOrErr
 
 		if poe.Err != nil {
```

Comment:
Why are drop all this? It's necessary to safely copy the various possible Err types.

Reply:
It was broken. The new `TestPackageTreeCopy` fails with the old logic - a `build.NoGoError`'s `Dir` field is not copied.

Why do we need to copy them? Wouldn't it be a mistake to mutate an error anyways?

Reply:
Alternatively, we could call `.Error()` and make a new error from the string. This is how the cache persists them.